amigo-lib/src/main/java/me/ele/amigo/AmigoInstrumentation.java

```diff
@@ -0,0 +1,217 @@
+package me.ele.amigo;
+
+import android.app.Activity;
+import android.app.Fragment;
+import android.app.Instrumentation;
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.content.pm.ActivityInfo;
+import android.os.Bundle;
+import android.os.IBinder;
+import android.os.UserHandle;
+import android.util.Log;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import me.ele.amigo.reflect.FieldUtils;
+import me.ele.amigo.utils.ComponentUtils;
+
+public class AmigoInstrumentation extends Instrumentation implements IInstrumentation {
+
+    public static final String EXTRA_TARGET_INTENT = "me.ele.amigo.OldIntent";
+    public static final String EXTRA_TARGET_INFO = "me.ele.amigo.OldInfo";
+    public static final String EXTRA_STUB_NAME = "me.ele.amigo.stub";
+
+    private static final String TAG = AmigoInstrumentation.class.getSimpleName();
+
+    private Instrumentation oldInstrumentation;
+
+    public AmigoInstrumentation(Instrumentation oldInstrumentation) {
+        this.oldInstrumentation = oldInstrumentation;
+    }
+
+    private void startStubActivity(Context who, Intent intent) {
+        ComponentName componentName = intent.getComponent();
+        ActivityStub.beforeStartActivity(getActivityInfo(who, componentName.getClassName()));
+        Class stubClazz = getDelegateActivityName(who, componentName.getClassName());
+        if (stubClazz == null) {
+            Log.e(TAG, "startStubActivity: weird, no stubs available for now.");
+            return;
+        }
+
+        Intent stubIntent = new Intent();
+        stubIntent.setComponent(new ComponentName(componentName.getPackageName(), stubClazz.getName()));
+        stubIntent.putExtra(EXTRA_TARGET_INTENT, intent);
+        stubIntent.setFlags(intent.getFlags());
+        intent.putExtra(EXTRA_STUB_NAME, stubClazz);
+        who.startActivity(stubIntent);  //TODO WE CAN AVOID THIS
+        ActivityStub.onActivityCreated(stubClazz, null, componentName.getClassName());
+    }
+
+    @Override
+    public ActivityResult execStartActivity(Context who, IBinder contextThread, IBinder token, Activity target,
+                                            Intent intent, int requestCode, Bundle options) {
+        try {
+            Method method = oldInstrumentation.getClass().getDeclaredMethod("execStartActivity", Context.class, IBinder.class, IBinder.class, Activity.class, Intent.class, int.class, Bundle.class);
+            return (ActivityResult) method.invoke(oldInstrumentation, who, contextThread, token, target, intent, requestCode, options);
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
+        } catch (IllegalAccessException e) {
+            e.printStackTrace();
+        } catch (InvocationTargetException e) {
+            e.printStackTrace();
+        }
```

Comment:
***API doc*** : If an exception is thrown during the invocation it is ***caught*** and
      wrapped in an InvocationTargetException. This exception is then thrown.